app/models/user.rb

```diff
@@ -0,0 +1,17 @@
+class User < ApplicationRecord
+  before_save :downcase_email
+  validates :name, presence: true,
+   length: {maximum: Settings.user.name.max_length}
+  validates :email, presence: true,
+    length: {maximum: Settings.user.email.max_length},
+    format: {with: Settings.user.email.valid_email_regex},
+    uniqueness: true
+  validates :password, presence: true,
+    length: {minimum: Settings.user.password.min_length}
+  has_secure_password
```

Comment:
dòng này để làm gì em?

Reply:
dòng trên của gem bcrypt mã hóa mật khẩu dựa trên các thuật toán mã hóa Blowfish thành chuỗi password_digest, đây là mã hóa một chiều và không thể lấy lại khi đã biết chuỗi mật khẩu trong database.
has_secure_password tự tạo ra 2 feild password và password_confirmation. Nó sẽ chỉ đọc password và đối chiều với trường password_confirmation để xác thực mật khẩu

Reply:
phần downcase:
downcase! thay đổi giá trị của biến còn downcase thì không thay đổi
downcase! trả về nil khi chuỗi không có kí tự viết hoa nào còn downcase luôn trả về một chuỗi viết thường